subprojects/core-api/src/main/java/org/gradle/api/artifacts/repositories/ArtifactResolutionDetails.java

```diff
@@ -0,0 +1,97 @@
+package org.gradle.api.artifacts.repositories;
+
+import org.gradle.api.Action;
+import org.gradle.api.Incubating;
+import org.gradle.api.artifacts.ModuleIdentifier;
+import org.gradle.api.artifacts.component.ModuleComponentIdentifier;
+import org.gradle.api.attributes.AttributeContainer;
+
+/**
+ * Details about an artifact resolution query. This is used whenever repository
+ * {@link ArtifactRepository#contentFilter(Action)} content filtering is in place.
+ * <p></p>
+ * This interface gives access to the details of the artifact query. There are two
+ * cases when filtering can be called:
+ * <p></p>
+ * <ul>
+ *     <li>when looking for a specific module version, for example org:foo:1.0</li>
+ *     <li>when looking for the list of versions for a module, for example org:foo</li>
+ * </ul>
+ * <p></p>
+ * Listing is called when using dynamic versions (ranges, 1.+, ...).
+ * <p></p>
+ * The module identifier will always be non-null. If what you want to express
+ * is that a module cannot be found in a repository, independently of its version,
+ * then it's enough to just look at the module id using {@link #getModuleId()}.
+ * <p></p>
+ * However, if you have to differentiate depending on the version number (for example,
+ * some versions of a module are found in one repository, others in a different repository),
+ * then you must look at the version thanks to the {@link #getComponentId()} method. But
+ * because there can also be version listings, you must also check for {@link #getModuleId()}.
+ * <p></p>
+ * A {@link #isVersionListing() convenience method} will make it easier to find out if you
+ * are in the version listing case, or module version case.
+ * <p></p>
+ * Filtering is done by calling the {@link #notFound()} method: as soon as you know a module
+ * cannot be found in a repository, call this method. Otherwise, Gradle will perform a request
+ * to find out. It doesn't matter if the module is eventually not found, as Gradle would handle
+ * this appropriately by looking at the next repository: the consequence is just a remote call.
+ *
+ * @since 5.1
+ */
+@Incubating
+public interface ArtifactResolutionDetails {
+    /**
+     * The identifier of the module being looked for in this repository
+     * @return the module identifier
+     */
+    ModuleIdentifier getModuleId();
+
+    /**
+     * The module component identifier of the module being looked for in this repository,
+     * which includes a version.
+     * @return the module version identifier. If it's a version listing, then this will
+     * be null.
+     */
+    ModuleComponentIdentifier getComponentId();
+
+    /**
+     * The attributes of the consumer looking for this module
+     * @return the consumer attributes
+     */
+    AttributeContainer getConsumerAttributes();
+
+    /**
+     * The name of the consumer. Usually corresponds to the name of the configuration being
+     * resolved.
+     * @return the consumer name
+     */
+    String getConsumerName();
```

Comment:
I'm also not sure we want to expose this, and it's left open for discussion. However during implementation I found that there are use cases that cannot be solved without this. Attributes will _not_ give all the necessary context about the consumer. In particular, "for runtime" usage doesn't tell you if it's "test runtime". Also attributes are not very user friendly, and mostly driven towards plugin authors. I therefore chose to make concessions here and expose it as a convenience to solve some use cases. This doesn't prevent us from deprecating later if we provide a better way to model it, but it seems to be an example of how being "pedantic" about how we should expose things or not is not helping the user to solve problems.

Reply:
Rather than adding speculatively and deprecating later, I'd rather not add this now, and instead reduce the ceremony around adding a custom attribute to all `test*` configurations.

Reply:
I think we should remove this and worry about that when we actually have a use case. I can think of none for "is this resolving for test runtime?".

Reply:
> reduce the ceremony around adding a custom attribute to all test* configurations

I'd like to avoid this as much as possible. Attributes are hard to grasp and hard to make right. We certainly do not want to see a proliferation of custom attributes. On the other hand `testCompileClasspath` is easy to figure out.